libraries/extractor/src/main/java/androidx/media3/extractor/ts/DtsUhdReader.java

```diff
@@ -0,0 +1,246 @@
+          }
+          break;
+        case STATE_READING_SAMPLE:
+          int bytesToRead = min(data.bytesLeft(), sampleSize - bytesRead);
+          output.sampleData(data, bytesToRead);
+          bytesRead += bytesToRead;
+          if (bytesRead == sampleSize) {
+            if (timeUs != C.TIME_UNSET) {
+              output.sampleMetadata(timeUs, C.BUFFER_FLAG_KEY_FRAME, sampleSize, 0, null);
+              timeUs += sampleDurationUs;
+            }
+            state = STATE_FINDING_SYNC;
+          }
+          break;
+        default:
+          throw new IllegalStateException();
+      }
+    }
+  }
+
+  @Override
+  public void packetFinished(boolean isEndOfInput) {
+    // Do nothing.
+  }
+
+  /**
+   * Continues a read from the provided {@code source} into a given {@code target}. It's assumed
+   * that the data should be written into {@code target} starting from an offset of zero.
+   *
+   * @param source The source from which to read.
+   * @param target The target into which data is to be read.
+   * @param targetLength The target length of the read.
+   * @return Whether the target length was reached.
+   */
+  private boolean continueRead(ParsableByteArray source, byte[] target, int targetLength) {
+    int bytesToRead = min(source.bytesLeft(), targetLength - bytesRead);
+    source.readBytes(target, bytesRead, bytesToRead);
+    bytesRead += bytesToRead;
+    return bytesRead == targetLength;
+  }
+
+  /**
+   * Locates the next SYNC value in the buffer, advancing the position to the byte that immediately
+   * follows it. If SYNC was not located, the position is advanced to the limit.
+   *
+   * @param pesBuffer The buffer whose position should be advanced.
+   * @return Whether SYNC was found.
+   */
+  private boolean skipToNextSync(ParsableByteArray pesBuffer) {
+    while (pesBuffer.bytesLeft() > 0) {
+      syncBytes <<= 8;
+      syncBytes |= pesBuffer.readUnsignedByte();
+      isFtocSync = DtsUtil.isUhdFtocSyncWord(syncBytes);
+      if (isFtocSync || DtsUtil.isUhdFtocNonSyncWord(syncBytes)) {
+        byte[] headerData = headerScratchBytes.getData();
+        headerData[0] = (byte) ((syncBytes >> 24) & 0xFF);
+        headerData[1] = (byte) ((syncBytes >> 16) & 0xFF);
+        headerData[2] = (byte) ((syncBytes >> 8) & 0xFF);
+        headerData[3] = (byte) (syncBytes & 0xFF);
+        bytesRead = 4;
+        syncBytes = 0;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Find the actual size of DTS UHD frame header. */
+  private void findHeaderSize() {
+    byte[] frameData = new byte[FTOC_HEADER_SIZE_MIN];
+    headerScratchBytes.setPosition(0);
+    headerScratchBytes.readBytes(frameData, 0, FTOC_HEADER_SIZE_MIN);
+    headerSizeToRead = DtsUtil.parseDtsUhdHeaderSize(frameData);
+    if (headerSizeToRead < bytesRead) { // Already read more data than the actual header size
+      headerSizeToRead = bytesRead; // Setting target read length equal to bytesRead
```

Comment:
Why is this needed? The test cases fail when this is removed for `DtsUhdReader` but still passes for `DtsReader`.

Reply:
For the DTS UHD stream, the header size is not constant. To find the actual header size, first, we need to parse a few bytes of data(FTOC_MIN_HEADER_SIZE). Sometimes, the actual header size may be less than the FTOC_MIN_HEADER_SIZE.
i.e.   bytesRead > headerSizeToRead
In this case, if we pass the actual 'headerSizeToRead' to the _continueRead()_ function invoked in the 'STATE_READING_HEADER' state, the _readBytes()_ will fail(the number of bytes to read will become negative).